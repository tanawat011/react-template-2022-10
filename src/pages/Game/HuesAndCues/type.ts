/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
