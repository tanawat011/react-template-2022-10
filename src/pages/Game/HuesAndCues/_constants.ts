export const SESSION = {
  ID: 'sessionId',
  ROOM_ID: 'roomId',
  TEMP_ID: 'tempSessionId',
}

export const COLLECTION = {
  HUES_AND_CUES: 'hues_and_cues',
  PLAYERS: 'players',
}

export const DOCUMENT = {
  ROOMS: 'rooms',
}

export const FIRESTORE_PATH = {
  ROOMS_DOCUMENT: `${COLLECTION.HUES_AND_CUES}/${DOCUMENT.ROOMS}`,
  PLAYERS_COLLECTION: `${COLLECTION.PLAYERS}`,
}
