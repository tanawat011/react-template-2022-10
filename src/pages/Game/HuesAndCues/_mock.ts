import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
