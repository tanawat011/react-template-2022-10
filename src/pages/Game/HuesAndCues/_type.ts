export type Player = {
  uid: string
  name: string
}

export type RoomPlayer = {
  uid: string
  name: string
  score: number
  color: string // * color is a hex value
  cells: string[] // * cells that the player has selected
  owner: boolean // * if true, the player is the owner of the room
  number: number // * sequential number of player in room
  hinter?: boolean // * is the player the hinter
  choice?: string[] // * choice is an array of colorIds that the player can select
  result?: string // * result is the colorId of the color that the player has selected
  isSubmitResult?: boolean // * If true, the player has responded to the hint
  isMyTurn?: boolean // * If true, it is the player's turn
  isStartGame?: boolean // * If true, the game has started
  countTurn?: number // * The number of turns the player has taken
  countRound?: number // * The number of rounds that have passed
}
