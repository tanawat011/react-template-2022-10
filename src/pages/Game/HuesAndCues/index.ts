export * from './HuesAndCues'
export * from './SetDisplayName'
export * from './SetupRoom'
