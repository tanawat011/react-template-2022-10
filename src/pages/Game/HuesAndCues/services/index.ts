export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
