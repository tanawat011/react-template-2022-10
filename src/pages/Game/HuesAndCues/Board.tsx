import type { RoomPlayer } from './_type'
import type { HuesAndCues } from 'data/HuesAndCues'

import { TwCol, TwRow } from 'components/Common'
import { getColTitles, getRowTitles, huesAndCues } from 'data/HuesAndCues'

import { CellClickable, CellTransparent, Marker } from './Common'
import { setPlayerInTheRoom } from './services'

const ColTitles = () => {
  const titles = ['', ...getColTitles(), '']

  return (
    <TwRow>
      {titles.map((title, index) => {
        return <CellTransparent key={`col-title-${title}-${index}`}>{title}</CellTransparent>
      })}
    </TwRow>
  )
}

const RowTitles = () => {
  const titles = getRowTitles()

  return (
    <TwCol>
      {titles.map((title, index) => {
        return <CellTransparent key={`row-title-${title}-${index}`}>{title}</CellTransparent>
      })}
    </TwCol>
  )
}

type Prop = {
  roomId: string
  currentPlayer: RoomPlayer
  playersInRoom: RoomPlayer[]
  updateCurrentPlayer: (currentPlayer: RoomPlayer) => Promise<void>
}

export const Board: React.FC<Prop> = ({
  roomId,
  currentPlayer,
  playersInRoom,
  updateCurrentPlayer,
}) => {
  const handlePlayerCells = async (item: HuesAndCues) => {
    const countTurn = currentPlayer?.countTurn || 0

    if (currentPlayer && !currentPlayer.hinter && currentPlayer.isMyTurn && countTurn < 1) {
      let cells = currentPlayer.cells
      const isDuplicateCell = cells.includes(item.id)

      if (isDuplicateCell) {
        cells = cells.filter((cell) => cell !== item.id)
      } else {
        cells.push(item.id)
      }

      // * In case has 2 players in the room
      if (playersInRoom.length === 2) {
        const _currentPlayer: RoomPlayer = {
          ...currentPlayer,
          cells,
          countTurn: countTurn + 1,
        }

        await updateCurrentPlayer(_currentPlayer)

        return
      }

      const playersInRoomSort = playersInRoom.sort((a, b) => a.number - b.number)

      let nextNumber = currentPlayer.number // 2
      let nextPlayerTurn = playersInRoomSort[nextNumber]

      // * In case the last player in the room
      if (!nextPlayerTurn) {
        nextNumber = 0
        nextPlayerTurn = playersInRoomSort[nextNumber]
      }

      // * In case the next player is the hinter
      if (nextPlayerTurn && nextPlayerTurn.hinter) {
        nextNumber += 1
        nextPlayerTurn = playersInRoomSort[nextNumber] || playersInRoomSort[0]
      }

      const _nextPlayerTurn: RoomPlayer = {
        ...nextPlayerTurn,
        isMyTurn: true,
      }

      await setPlayerInTheRoom(roomId, _nextPlayerTurn)

      const _currentPlayer: RoomPlayer = {
        ...currentPlayer,
        cells,
        isMyTurn: false,
        countTurn: countTurn + 1,
      }

      await updateCurrentPlayer(_currentPlayer)
    }
  }

  return (
    <>
      <ColTitles />
      <TwRow>
        <RowTitles />

        <TwCol>
          {huesAndCues.map((items, index) => {
            return (
              <TwRow key={`hues-and-cues-${index}`}>
                {items.map((cell) => {
                  const cellId = cell.id
                  const cellColor = cell.color

                  const player = playersInRoom.find((u) => u.cells.includes(cellId))

                  const hinter = playersInRoom.find((u) => u.hinter && !!u.result)
                  const isHint = hinter?.result === cellId

                  const isShowResultForHinter = currentPlayer.hinter
                  const isShowResultForPlayer = !currentPlayer.hinter && hinter?.isSubmitResult
                  const isShowResult = isHint && (isShowResultForHinter || isShowResultForPlayer)

                  return (
                    <CellClickable
                      key={`${cellColor}-${cellId}`}
                      className={cellColor}
                      onClick={() => handlePlayerCells(cell)}
                    >
                      {player?.color && <Marker className={player.color} />}
                      {isShowResult && (
                        <div className='absolute w-28 h-28 border-4 border-solid border-white' />
                      )}
                    </CellClickable>
                  )
                })}
              </TwRow>
            )
          })}
        </TwCol>

        <RowTitles />
      </TwRow>
      <ColTitles />
    </>
  )
}
