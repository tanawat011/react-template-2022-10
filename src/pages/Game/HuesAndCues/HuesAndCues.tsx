import type { RoomPlayer } from './_type'
import type { Unsubscribe } from 'firebase/firestore'

import { useEffect, useState } from 'react'

import { useNavigate, useParams } from 'react-router-dom'
import { v4 as uuid } from 'uuid'

import { TwCol } from 'components/Common'
import { useSessionStorage } from 'hooks'

import { Board } from './Board'
import { ChooseColorHint } from './ChooseColorHint'
import { ChooseColorToken } from './ChooseColorToken'
import { NextHinter } from './NextHinter'
import { NextRound } from './NextRound'
import { PlayerColor } from './PlayerColor'
import { RemoveRoom } from './RemoveRoom'
import { ResetRoom } from './ResetRoom'
import { ScoreBoard } from './ScoreBoard'
import { StartGame } from './StartGame'
import { SubmitResult } from './SubmitResult'
import { SESSION } from './_constants'
import {
  getAllPlayersInTheRoom,
  getPlayerInTheRoom,
  setPlayerInTheRoom,
  subscribeRoom,
  toSetDisplayRoom,
} from './services'

export const HuesAndCues = () => {
  const urlParam = useParams<{ roomId: string }>()
  const navigate = useNavigate()

  const [sessionId, , removeSessionId] = useSessionStorage(SESSION.ID)
  const [roomId, setRoomId, removeRoomId] = useSessionStorage(SESSION.ROOM_ID)
  const [tempSessionId, setTempSessionId, removeTempSessionId] = useSessionStorage(SESSION.TEMP_ID)

  const [playersInRoom, setPlayersInRoom] = useState<RoomPlayer[]>([])
  const [currentPlayer, setCurrentPlayer] = useState<RoomPlayer>()

  let unsubscribe: Unsubscribe = () => {
    return
  }

  useEffect(() => {
    checkRoomAndFetchData()

    return () => {
      unsubscribe()
    }
  }, [])

  const checkRoomAndFetchData = async () => {
    const _roomId = urlParam.roomId || ''

    const _playersInRoom = await getAllPlayersInTheRoom(_roomId)

    if (!_playersInRoom.length) {
      removeSessionId()
      removeRoomId()
      removeTempSessionId()

      toSetDisplayRoom(navigate)

      return
    }

    if (!sessionId) {
      const _sessionId = tempSessionId || uuid()

      setRoomId(_roomId)
      setTempSessionId(_sessionId)

      toSetDisplayRoom(navigate)

      return
    }

    await fetchCurrentPlayer()

    fetchPlayersInRoom()
  }

  const fetchCurrentPlayer = async () => {
    const data = await getPlayerInTheRoom(roomId, sessionId)

    setCurrentPlayer(data)
  }

  const updateCurrentPlayer = async (_currentPlayer: RoomPlayer) => {
    setCurrentPlayer(_currentPlayer)

    await setPlayerInTheRoom(roomId, _currentPlayer)
  }

  const fetchPlayersInRoom = () => {
    const callback = (data: RoomPlayer[]) => {
      setPlayersInRoom(data)
      fetchCurrentPlayer()
    }

    unsubscribe = subscribeRoom(roomId, callback)
  }

  const isOwnerRoom = currentPlayer && currentPlayer.owner
  const isStartGame = playersInRoom.findIndex((p) => p.isStartGame) > -1
  const isHinter = currentPlayer && currentPlayer.hinter
  const isPlayerNoColor = currentPlayer && !currentPlayer.color
  const isPlayerHasColor = currentPlayer && currentPlayer.color

  return (
    <TwCol className='p-6'>
      {isOwnerRoom && (
        <>
          <RemoveRoom roomId={roomId} />

          <ResetRoom
            roomId={roomId}
            currentPlayer={currentPlayer}
            setCurrentPlayer={setCurrentPlayer}
          />

          <StartGame
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentPlayer={updateCurrentPlayer}
          />
        </>
      )}

      {currentPlayer?.hinter && (
        <>
          <NextHinter
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentHinter={updateCurrentPlayer}
          />

          <SubmitResult
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentPlayer={updateCurrentPlayer}
          />

          <NextRound
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            setCurrentPlayer={setCurrentPlayer}
          />
        </>
      )}

      <ScoreBoard playersInRoom={playersInRoom} />

      {isPlayerNoColor && (
        <ChooseColorToken
          currentPlayer={currentPlayer}
          playersInRoom={playersInRoom}
          updateCurrentPlayer={updateCurrentPlayer}
        />
      )}

      {isPlayerHasColor && (
        <>
          <PlayerColor
            currentPlayer={currentPlayer}
            playersInRoom={playersInRoom}
            updateCurrentPlayer={updateCurrentPlayer}
          />

          {isHinter && (
            <ChooseColorHint
              currentPlayer={currentPlayer}
              updateCurrentPlayer={updateCurrentPlayer}
            />
          )}

          {isStartGame && (
            <Board
              roomId={roomId}
              currentPlayer={currentPlayer}
              playersInRoom={playersInRoom}
              updateCurrentPlayer={updateCurrentPlayer}
            />
          )}
        </>
      )}
    </TwCol>
  )
}
