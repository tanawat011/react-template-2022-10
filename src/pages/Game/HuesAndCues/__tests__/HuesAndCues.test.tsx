import type { RoomPlayer } from '../_type'
import type { RenderResult } from '@testing-library/react'

import { fireEvent, screen, act, waitFor } from '@testing-library/react'
import { mocked } from 'jest-mock'

import { renderWithProviders } from 'helpers/test'

import { HuesAndCues } from '../HuesAndCues'
import { MOCK } from '../_mock'
import { getAllPlayersInTheRoom, getPlayerInTheRoom, subscribeRoom } from '../services'

jest.mock('../Board', () => ({
  Board: () => <div>Board</div>,
}))

jest.mock('../ChooseColorHint', () => ({
  ChooseColorHint: () => <div>ChooseColorHint</div>,
}))

jest.mock('../ChooseColorToken', () => ({
  ChooseColorToken: () => <div>ChooseColorToken</div>,
}))

jest.mock('../NextHinter', () => ({
  NextHinter: () => <div>NextHinter</div>,
}))

jest.mock('../NextRound', () => ({
  NextRound: () => <div>NextRound</div>,
}))

jest.mock('../PlayerColor', () => ({
  PlayerColor: () => <div>PlayerColor</div>,
}))

jest.mock('../RemoveRoom', () => ({
  RemoveRoom: () => <div>RemoveRoom</div>,
}))

jest.mock('../ResetRoom', () => ({
  ResetRoom: () => <div>ResetRoom</div>,
}))

jest.mock('../ScoreBoard', () => ({
  ScoreBoard: () => <div>ScoreBoard</div>,
}))

jest.mock('../StartGame', () => ({
  StartGame: ({ updateCurrentPlayer }: { updateCurrentPlayer: (payload: RoomPlayer) => void }) => (
    <div data-testid='button' onClick={() => updateCurrentPlayer(MOCK.THREE_PLAYER[0])}>
      StartGame
    </div>
  ),
}))

jest.mock('../SubmitResult', () => ({
  SubmitResult: () => <div>SubmitResult</div>,
}))

jest.mock('../services', () => ({
  getAllPlayersInTheRoom: jest.fn(),
  getPlayerInTheRoom: jest.fn(),
  setPlayerInTheRoom: jest.fn(),
  subscribeRoom: jest.fn(),
  toSetDisplayRoom: jest.fn(),
}))

describe('<HuesAndCues />', () => {
  test('match snapshot', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    mocked(getAllPlayersInTheRoom).mockResolvedValue(MOCK.THREE_PLAYER)
    mocked(getPlayerInTheRoom).mockResolvedValue(MOCK.THREE_PLAYER[0])
    mocked(subscribeRoom).mockImplementation(
      (_: unknown, callback: (payload: RoomPlayer[]) => void) => {
        callback(MOCK.THREE_PLAYER)

        return jest.fn()
      },
    )

    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<HuesAndCues />)
    })

    await waitFor(() => {
      expect(el.asFragment()).toMatchSnapshot()
    })
  })

  test('no one players in the room', async () => {
    mocked(getAllPlayersInTheRoom).mockResolvedValue([])
    mocked(subscribeRoom).mockImplementation(
      (_: unknown, callback: (payload: RoomPlayer[]) => void) => {
        callback(MOCK.THREE_PLAYER)

        return jest.fn()
      },
    )

    const { asFragment } = renderWithProviders(<HuesAndCues />)

    expect(asFragment()).toBeTruthy()
  })

  test('no session id', async () => {
    mocked(getAllPlayersInTheRoom).mockResolvedValue(MOCK.THREE_PLAYER)
    mocked(getPlayerInTheRoom).mockResolvedValue(MOCK.THREE_PLAYER[0])
    mocked(subscribeRoom).mockImplementation(
      (_: unknown, callback: (payload: RoomPlayer[]) => void) => {
        callback(MOCK.THREE_PLAYER)

        return jest.fn()
      },
    )

    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<HuesAndCues />)
    })

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('the owner clicks start game', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    mocked(getAllPlayersInTheRoom).mockResolvedValue(MOCK.PLAYER_HAS_COLOR)
    mocked(getPlayerInTheRoom).mockResolvedValue(MOCK.PLAYER_HAS_COLOR[0])
    mocked(subscribeRoom).mockImplementation(
      (_: unknown, callback: (payload: RoomPlayer[]) => void) => {
        callback(MOCK.PLAYER_HAS_COLOR)

        return jest.fn()
      },
    )

    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<HuesAndCues />)
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('game is started', async () => {
    sessionStorage.setItem('sessionId', 'session-id-x')
    mocked(getAllPlayersInTheRoom).mockResolvedValue(MOCK.PLAYER_HAS_COLOR_START_GAME)
    mocked(getPlayerInTheRoom).mockResolvedValue(MOCK.PLAYER_HAS_COLOR_START_GAME[0])
    mocked(subscribeRoom).mockImplementation(
      (_: unknown, callback: (payload: RoomPlayer[]) => void) => {
        callback(MOCK.PLAYER_HAS_COLOR_START_GAME)

        return jest.fn()
      },
    )

    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(<HuesAndCues />)
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })
})
