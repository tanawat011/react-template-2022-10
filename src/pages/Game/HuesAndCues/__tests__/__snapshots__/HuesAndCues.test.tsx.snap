// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HuesAndCues /> match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0 p-6"
  >
    <div>
      RemoveRoom
    </div>
    <div>
      ResetRoom
    </div>
    <div
      data-testid="button"
    >
      StartGame
    </div>
    <div>
      NextHinter
    </div>
    <div>
      SubmitResult
    </div>
    <div>
      NextRound
    </div>
    <div>
      ScoreBoard
    </div>
    <div>
      ChooseColorToken
    </div>
  </div>
</DocumentFragment>
`;
